---
layout: page
title: About
---

We developed a deconvolution method, 'IRIS', that can characterize the spatial organization of complex tissues through accurate and efficient detection of spatial domains. IRIS is unique in its ability in leveraging single-cell RNA-seq data for reference-informed spatial domain detection, integrating multiple SRT tissue slices jointly while explicitly accounting for the correlation both within and across slices, and taking advantage of multiple algorithmic innovations for highly scalable computation. We demonstrate the advantages of IRIS through in-depth analysis of six SRT datasets from different technologies across distinct tissues, species, and spatial resolutions. In addition, IRIS comes with an efficient optimization framework with multiple algebraic innovations for scalable computation and can easily handle multiple spatial transcriptomics datasets with millions of spatial locations and tens of thousands of genes

Cite `IRIS`
-------------------
Ma Y, Zhou X. Accurate and efficient integrative reference-informed spatial domain detection for spatial transcriptomics. Nat Methods. 2024 Jun 6. doi: 10.1038/s41592-024-02284-9. Epub ahead of print. PMID: 38844627.

Contact
-------------------
if you have questions, feel free to leave messages on the [github issues](https://github.com/YingMa1993/IRIS/issues) 



